### PR TITLE
Refactor visualization functions

### PR DESCRIPTION
**Pull request by jofrevalles**
### Summary
Plot recipes currently describe plot primitives that live in an `Axis` but can not change the `Axis` parameters. Therefore, with plot recipes, we can't currently include `Colorbar`s, add interactions, and in general, customize the `Axis`.
In this PL we address #27, #25, and we change the `plot` function so now it can take a `GridPosition` and plot into it. This enabled us to add the following features:
- Added integration to `plot` the contraction paths into subplots.
- Added the possibility to create 3D plots.
- Grids and decorations are removed by default in 2D plots.
- Two colorbars have been added that indicate the corresponding values of `flops` and `size`.

### Example
```julia
julia> using OptimizedEinsum
julia> using Makie
julia> using GLMakie
julia> output, inputs, size_dict = rand_equation(15, 4)
(Symbol[], [[:f, :e, :l, :s, :q], [:v, :D, :e, :u, :q, :i], [:w, :a, :y], [:C, :h, :A, :d, :c, :y, :m], [:x, :u], [:s, :g, :a, :d, :f, :l, :t], [:z, :C, :k, :j, :p, :b], [:r, :b], [:v, :m], [:D], [:n, :B], [:h, :w, :z, :t, :p], [:i, :k, :o], [:B, :n, :x], [:g, :r, :o, :A, :c, :j]], Dict(:o => 7, :b => 2, :p => 5, :n => 9, :j => 4, :e => 4, :D => 5, :B => 5, :c => 2, :h => 6…))
julia> path = contractpath(RandomGreedy, inputs, output, size_dict)
ContractionPath([(1, 6), (4, 9), (8, 15), (7, 12), (11, 14), (18, 13), (2, 10), (20, 5), (23, 22), (24, 16), (17, 3), (26, 19), (27, 21), (25, 28)], [[:f, :e, :l, :s, :q], [:v, :D, :e, :u, :q, :i], [:w, :a, :y], [:C, :h, :A, :d, :c, :y, :m], [:x, :u], [:s, :g, :a, :d, :f, :l, :t], [:z, :C, :k, :j, :p, :b], [:r, :b], [:v, :m], [:D], [:n, :B], [:h, :w, :z, :t, :p], [:i, :k, :o], [:B, :n, :x], [:g, :r, :o, :A, :c, :j]], Symbol[], Dict(:o => 7, :b => 2, :p => 5, :n => 9, :j => 4, :e => 4, :D => 5, :B => 5, :c => 2, :h => 6…))
julia> f = Figure(backgroundcolor="ivory", resolution=(1000, 700))
julia> scatter(f[1,1], 1..4, log)
Makie.AxisPlot(Axis (1 plots), Scatter{Tuple{Vector{Point{2, Float32}}}})
julia> lines(f[2,1], 1..10, cos)
Makie.AxisPlot(Axis (1 plots), Lines{Tuple{Vector{Point{2, Float32}}}})
julia> plot!(f[1:2,2:3], path)
Scene (1000px, 700px):
  0 Plots
  3 Child Scenes:
    ├ Scene (1000px, 700px)
    ├ Scene (1000px, 700px)
    └ Scene (1000px, 700px)
GridPosition(GridLayout[2, 3] (3 children), GridLayoutBase.Span(1:2, 2:3), GridLayoutBase.Inner())
```
image

And we can also plot in 3D with:
```julia
julia> using NetworkLayout
julia> plot(path; layout=Spring(dim=3))
```
image

### CODE REVIEW

**Comment by mofeing** [comment]
In the first plot, is the path plot a 3D plot?

**Comment by jofrevalles** [comment]
> In the first plot, is the path plot a 3D plot?
Oh sorry, it was not, let me correct the code example.

File: src/Visualization.jl [hunk 1]
```diff
@@ -2,33 +2,105 @@ using Graphs
...
+    ax = current_axis(current_figure())
+
+    if !isa(ax, LScene) # hide decorations if it is not a 3D plot
+        hidedecorations!(ax)
+        hidespines!(ax)
+        ax.aspect = DataAspect()
     end
 
-    graphplot!(P, graph; kwargs...)
+    cbar = Colorbar(f[1,2], get_edge_plot(p), label=L"\log_{2}(size)", flip_vertical_label=true, labelsize = 34)
```
Comment (mofeing) [inline src/Visualization.jl:96]: Remove the LatexStrings dependency.

```suggestion
    cbar = Colorbar(f[1,2], get_edge_plot(p), label="\\log\_\{2\}(size)", flip_vertical_label=true, labelsize = 34)
```

File: src/Visualization.jl [hunk 2]
```diff
@@ -2,33 +2,105 @@ using Graphs
...
+    cbar.height = Relative(5/6)
+
+    cbar2 = Colorbar(f[1,0], get_node_plot(p), label=L"\log_{10}(flops)", flipaxis=false, labelsize = 34)
```
Comment (mofeing) [inline src/Visualization.jl:99]:
```suggestion
    cbar2 = Colorbar(f[1,0], get_node_plot(p), label="\\log\_\{10\}(flops)", flipaxis=false, labelsize = 34)
```

Comment (mofeing) [reply src/Visualization.jl:99]: You may need to use `raw"..."` strings.

File: src/Visualization.jl [hunk 3]
```diff
@@ -2,33 +2,105 @@ using Graphs
...
+
+    p = graphplot!(f[1,1], graph;
+    arrow_attr = (colorrange=(min_size, max_size), colormap=colormap),
+    edge_attr = (colorrange=(min_size, max_size), colormap=colormap),
+    node_attr = (colorrange=(min_flop, max_flop),
+    colormap = to_colormap(:plasma)[begin:end-50]), kwargs...)
+
+    ax = current_axis(current_figure())
+
+    if !isa(ax, LScene) # hide decorations if it is not a 3D plot
```
Comment (mofeing) [inline src/Visualization.jl:90]:
```suggestion
    # hide decorations if it is not a 3D plot
    if !isa(ax, LScene)
```

File: src/Visualization.jl [hunk 4]
```diff
@@ -2,33 +2,105 @@ using Graphs
...
 
-    graphplot!(P, graph; kwargs...)
+    cbar = Colorbar(f[1,2], get_edge_plot(p), label=L"\log_{2}(size)", flip_vertical_label=true, labelsize = 34)
+    cbar.height = Relative(5/6)
+
+    cbar2 = Colorbar(f[1,0], get_node_plot(p), label=L"\log_{10}(flops)", flipaxis=false, labelsize = 34)
+    cbar2.height = Relative(5/6)
+
+    cbar2.alignmode = Mixed(left = -10, right = -30)
+    cbar.alignmode = Mixed(left = -30, right = -10)
```
Comment (mofeing) [inline src/Visualization.jl:103]:
```suggestion
    # TODO configurable alignments
    cbar2.alignmode = Mixed(left = -10, right = -30)
    cbar.alignmode = Mixed(left = -30, right = -10)
```

File: src/Visualization.jl (hunk 1 above)
Comment (mofeing) [inline src/Visualization.jl:96]:
```suggestion
    # TODO configurable `labelsize`
    cbar = Colorbar(f[1,2], get_edge_plot(p), label=L"\log_{2}(size)", flip_vertical_label=true, labelsize = 34)
```

**Review by mofeing — CHANGES_REQUESTED** [review]

File: src/Visualization.jl [hunk 5]
```diff
@@ -2,33 +2,105 @@ using Graphs
...
+    min_size, max_size = extrema(log_size)
+    min_flop, max_flop = extrema(log_flop)
+
+    kwargs[:edge_width] = (log_size/max_size)*MAX_EDGE_WIDTH
+    kwargs[:arrow_size] = (log_size/max_size)*MAX_ARROW_SIZE
+    kwargs[:edge_color] = log_size
+    kwargs[:node_size] = (log_flop/max_flop)*MAX_NODE_SIZE
+    kwargs[:node_color] = log_flop
+
+    haskey(kwargs, :layout) && isa(kwargs[:layout],Spring{3}) ? ax = LScene(f[1,1]) : ax = Axis(f[1,1])
```
[the remark this answers is not in the document]
Comment (jofrevalles) [reply src/Visualization.jl:80]: Yes... I have to look into this

[the remark this answers is not in the document]
Comment (mofeing) [reply src/Visualization.jl:81]: btw, you can use `isa` as an infix operator like `kwargs[:layout] isa Spring{3}`

**Comment by mofeing** [comment]
@jofrevalles I think it is failing in Code Coverage. can you check it out tomorrow?

File: src/Visualization.jl [hunk 6]
```diff
@@ -2,33 +2,110 @@ using Graphs
...
+    # hide decorations if it is not a 3D plot
+    if !isa(ax, LScene)
+        hidedecorations!(ax)
+        hidespines!(ax)
+        ax.aspect = DataAspect()
     end
 
-    graphplot!(P, graph; kwargs...)
+    # TODO configurable `labelsize`
+    cbar = Colorbar(f[1,2], get_edge_plot(p), label="\\log\_\{2\}(size)", flip_vertical_label=true, labelsize = 34)
```
Comment (mofeing) [inline src/Visualization.jl:100]:
```suggestion
    cbar = Colorbar(f[1,2], get_edge_plot(p), label=raw"\log_{2}(size)", flip_vertical_label=true, labelsize = 34)
```

File: src/Visualization.jl [hunk 7]
```diff
@@ -2,33 +2,110 @@ using Graphs
...
+    cbar.height = Relative(5/6)
+
+    cbar2 = Colorbar(f[1,0], get_node_plot(p), label="\\log\_\{10\}(flops)", flipaxis=false, labelsize = 34)
```
Comment (mofeing) [inline src/Visualization.jl:103]:
```suggestion
    cbar2 = Colorbar(f[1,0], get_node_plot(p), label=raw"\log_{10}(flops)", flipaxis=false, labelsize = 34)
```

**Review by mofeing — CHANGES_REQUESTED** [review]
I think these changes should fix the CI workflow. What I'm not sure is if we need to wrap everything between $ symbols like in LaTeX for equations.

File: src/Visualization.jl [hunk 8]
```diff
@@ -1,34 +1,72 @@
...
+    haskey(kwargs, :layout) && kwargs[:layout] isa Spring{3} ? ax = LScene(f[1,1]) : ax = Axis(f[1,1])
+
+    p = graphplot!(f[1,1], graph;
+        arrow_attr = (colorrange=(min_size, max_size), colormap=colormap),
+        edge_attr = (colorrange=(min_size, max_size), colormap=colormap),
+        node_attr = (colorrange=(min_flop, max_flop),
+        # TODO replace `to_colormap(:plasma)[begin:end-50]), kwargs...)` with a custom colormap
+        colormap = to_colormap(:plasma)[begin:end-50]), kwargs...)
+
+    ax = current_axis(current_figure())
```
Comment (mofeing) [inline src/Visualization.jl:51]: Isn't `ax` defined just a couple of lines above?

File: src/Visualization.jl [hunk 9]
```diff
@@ -1,34 +1,72 @@
...
+
+    # hide decorations if it is not a 3D plot
+    if !isa(ax, LScene)
+        hidedecorations!(ax)
+        hidespines!(ax)
+        ax.aspect = DataAspect()
     end
```
Comment (mofeing) [inline src/Visualization.jl:58]: Integrate these lines with line 42?

**Review by mofeing — APPROVED** [review]

**Pull request merged**